azure/services/privatedns/privatedns.go

```diff
@@ -0,0 +1,131 @@
+		s.scope.ManagementClusterSubscriptionID(), // label: subscription_id
+	).Set(1)
+
+	privateClusterRecordSets, err := s.privateDNSClient.ListPrivateRecordSets(ctx, managementClusterResourceGroup, clusterZoneName)
+	if err != nil && !azure.IsParentResourceNotFound(err) {
+		return microerror.Mask(err)
+	} else if azure.IsParentResourceNotFound(err) {
+		log.V(1).Info("cluster specific private DNS zone not found, creating a new one")
+		err = s.privateDNSClient.CreateOrUpdatePrivateZone(ctx, managementClusterResourceGroup, clusterZoneName, armprivatedns.PrivateZone{
+			Name:     &clusterZoneName,
+			Location: to.StringPtr(capzazure.Global),
+		})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	log.Info("list virtualNetworkLinks")
+	networkLinks, err := s.privateDNSClient.ListVirtualNetworkLink(ctx, managementClusterResourceGroup, clusterZoneName)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	log.V(1).Info("list of all network links", "virtualNetworkLinks", networkLinks)
+
+	operatorGeneratedVirtualNetworkLinkIndex := slices.IndexFunc(networkLinks, func(virtualNetworkLink *armprivatedns.VirtualNetworkLink) bool {
+		return *virtualNetworkLink.Name == *to.StringPtr(s.scope.ClusterName() + "-dns-" + managementClusterResourceGroup + "-vnet-link")
+	})
+
+	if operatorGeneratedVirtualNetworkLinkIndex == -1 {
+		log.V(1).Info("virtual network link not found, creating a new one")
+
+		err = s.privateDNSClient.CreateOrUpdateVirtualNetworkLink(
+			ctx,
+			managementClusterResourceGroup,
+			clusterZoneName,
+			s.scope.ClusterName(),
+			s.scope.ManagementClusterVnetID(),
+		)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	log.V(1).Info("get privateDNSZone Object", "privateDNSZone", clusterZoneName)
+	privateZones, err := s.privateDNSClient.GetPrivateZone(ctx, s.scope.ManagementClusterResourceGroup(), clusterZoneName)
+	if err != nil {
+		log.V(1).Info("new error", "error", err.Error())
+	}
+
+	// dns_operator_zone_records_sum
+	metrics.ClusterZoneRecords.WithLabelValues(
+		clusterZoneName,
+		metrics.ZoneTypePrivate,
+	).Set(float64(to.Int64(privateZones.Properties.NumberOfRecordSets)))
+
+	log.V(1).Info("current known private Zones in management cluster", "privateZones", privateZones)
+
+	if err := s.updateARecords(ctx, privateClusterRecordSets); err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (s *Service) ReconcileDelete(ctx context.Context) error {
+	log := log.FromContext(ctx).WithName("azure-private-dns-delete")
+	clusterZoneName := s.scope.ClusterDomain()
+	log.Info("Reconcile DNS deletion", "privateDNSZone", clusterZoneName)
+
+	if err := s.privateDNSClient.DeleteVirtualNetworkLink(ctx, s.scope.ManagementClusterResourceGroup(), clusterZoneName, s.scope.ClusterName()); err != nil {
+		return microerror.Mask(err)
+	}
+
+	if err := s.privateDNSClient.DeletePrivateZone(ctx, s.scope.ManagementClusterResourceGroup(), clusterZoneName); err != nil {
+		return microerror.Mask(err)
+	}
```

Comment:
note to myself (@bavarianbidi ): on the first `reconcileDelete`, the `DeleteVirtualNetworkLink` already returned but the resource "might" be still there so the `DeletePrivateZone` will fail. Only after the second attempt, `DeletePrivateZone` succeed.

This is visible in the `apierror` metric